README.md
# JoshieHTTP
A webserver designed and programmed by a bored 15 year old that has nothing better to do.

# Table of contents 
[Running](#running)

[Adding sites](#creating-new-domains-in-mainconf)

[Proxying](#proxying-a-site-or-internal-site)

[Scripting](#scripting)

[SSL](#ssl)

# Running
Create the directory /var/www/html, or edit the `'default'` configuration location to point to another folder that already exists.

Create `index.html` in that folder and put whatever you want in it.

Run `./start.sh`

# Creating new domains in main.conf
The configuration file is a JSON file that contains the settings for every website you have running. If it's not found in there, the `'default'` configuration is used.

To create a new site, append a comma to the end of the last entry in the `main.conf`, in between the curly brackets.

In quotes, type out the full domain name, with separate entries for with and without a `www.`

After the quote, type a `:` and then an open curly bracket.

In quotes, type `"type"`, then a `:` and then in quotes `"local"`

Append a comma, then type in quotes `"location"`, a `:`, and then in quotes the folder you wish the domain to point to.

Ex:

```JSON
{
  "default": {"type": "local", "location": "/var/www/html"},
  "www.example.com": {"type": "local", "location": "/var/www/html2"}
}
```

# Proxying a site or internal site
JoshieHTTPD supports proxying of external or internal sites, and ports.

Follow the above instructions to add a new site, but instead of the "type" being "local", change "local" to "proxy"

Then, instead of a folder for the location, put the URL of the place you would like to proxy to. _make sure that you put a http:// or a https:// in front of the URL_. If you would like to proxy a port, simply append a : and then the port in side of the quotes.

Ex:

```JSON
{
  "default": {"type": "proxy", "location": "http://www.example.com"},
  "api.example.com": {"type": "proxy", "location": "http://localhost:8080"}
}
```

# Scripting
PHP, Ruby, Python, etc. is not supported in JoshieHTTP. Mainly because I didn't feel like adding them. Too much. So, to make up for that, I made the `.sjs` extension for scripting. It's just a .js file, but renamed. It runs in node.js, and all URL parameters and cookies (cookies in a single variable called "COOKIES") are passed on as environment variables to the script. You can go on from there, reading the variables. It's all run server-side, and to send data back you just send it back in a `console.log();`. There's actually a lot you can do with this. Happy scripting!

Update: PHP support now added. Some (_most_) features don't work, since it's just called directly from the command line, but it's a start! Just make sure PHP is installed on your server. Just add ```"php_enabled"``` to your configuration file for the sites you want PHP on. Happy scripting!

_Coming soon:
JoshieScript in the .jspt format_

# SSL
If you wish for JoshieHTTP to run with SSL (who wouldn't?), create a directory named `ssl` in the same directory as `index.js`

In `ssl`, you need two files. `key.pem`, which is your private key, and `cert.pem`, which is your certificate file.

Then, run `./start-ssl.sh`


changelog.md
# Changelog
2.0 - Initial creation (there was stuff before this)

Raw format download (fixes file encoding breaking stuff)

SSL added

.sjs added

Configuration file stuff

etc.

2.1 - Addition of 'POST' method for .sjs

Security stuff

Fixed stuff involving URL parameters and .sjs files

etc.

2.2 - Addition of PHP support

That's it

2.3 - Linux security enhancements

Made it switch to an unprivileged user after starting. That's it.

2.3.1 - Fixed some issues regarding URL encoding

Fixed some issues where I accidentally made it parse URLs preemptively before running sjs files.

3.0.0 - Changed how arguments are passed to .sjs files

Changed it so arguments are passed through environment variables instead of command line parameters, and allowed cookies to be passed to the script.

3.0.1 - Fixed potential exploit involving PATH environment variable in sjs, now sjs files execute in the working directory they're located in.

3.0.2 - Finally decided to pass the HTTP headers to .sjs files and proxied sites.

3.0.3 - Performance updates and fixed some error handling things

Stopped trying to be better than the built in URL constructer and no longer parses URLs manually.

Made sure that when you have a 404.html page, it doesn't just crash and not serve your 404.html file.

Also, added the ability to add a 500.html page to the root of your website directory and show something when a server error ocurrs!


webinstall.sh
#!/bin/sh

UserExists()
{
	awk -F":" '{ print $1 }' /etc/passwd | grep -x $1 > /dev/null
	return $2
}

if [ -d "~/.itmp" ]
then
	echo " * Removing installer tmp directory"
	rm -rf ~/.itmp
	echo " * Creating installer tmp directory"
	mkdir ~/.itmp
else
	echo " * Creating installer tmp directory"
	mkdir ~/.itmp
fi
echo " * Downloading server.js"
wget -O ~/.itmp/server.js http://joshiehttp.gq/dl/linux/server.js
echo " * Downloading default configuration"
wget -O ~/.itmp/main.conf http://joshiehttp.gq/dl/linux/main.conf
echo " * Downloading run files"
wget -O ~/.itmp/start.sh http://joshiehttp.gq/dl/linux/start.sh
wget -O ~/.itmp/start-ssl.sh http://joshiehttp.gq/dl/linux/start-ssl.sh
if [ ! -d "/var/www" ]
then
	echo " * Creating /var/www"
	mkdir /var/www
fi
if [ ! -d "/var/www/html" ]
then
	echo " * Creating /var/www/html"
	mkdir /var/www/html
fi
if [ ! -d "/var/www/log" ]
then 
	echo " * Creating log directory"
	mkdir /var/www/log
fi
echo " * Moving files"
mv -f ~/.itmp/* /var/www
echo " * chmodding files"
chmod 660 /var/www/main.conf
chmod 555 /var/www/server.js
chmod 660 /var/www
chmod 660 /var/www/log
chmod 550 /var/www/start.sh
chmod 550 /var/www/start-ssl.sh
UserExists secureweb
if [ $2 = 0 ]
then

else
	echo " * Creating 'secureweb' user"
	groupadd secureweb
	useradd -M -s /bin/sh -G secureweb secureweb
fi
echo " * chowning files"
chown -R secureweb:secureweb /var/www/html
chown -R secureweb:secureweb /var/www/log
echo " * Deleting temporary files"
rm -rf ~/.itmp
echo " * Done!"
echo "Go to /var/www and run ./start.sh (or ./start-ssl.sh if you've already set up ssl) to start the webserver! screen is recommended."


localinstall.sh
#!/bin/sh

UserExists()
{
	awk -F":" '{ print $1 }' /etc/passwd | grep -x $1 > /dev/null
	return $2
}

if [ -d "/var/www" ]
then
	echo " * Creating /var/www"
	mkdir /var/www
fi
if [ -d "/var/www/html" ]
then
	echo " * Creating /var/www/html"
	mkdir /var/www/html
fi
echo " * Moving files"
mv -f * /var/www/*
echo " * chmodding files"
chmod 660 /var/www/main.conf
chmod 555 /var/www/server.js
chmod 660 /var/www
chmod 550 /var/www/start.sh
chmod 550 /var/www/start-ssl.sh
UserExists secureweb
if [ $2 = 0 ]
then

else
	echo " * Creating 'secureweb' user"
	groupadd secureweb
	useradd -M -s /bin/sh -G secureweb secureweb
fi
echo " * chowning files"
chown -R secureweb:secureweb /var/www/html
echo " * Deleting temporary files"
rm -rf ~/.itmp
echo " * Done!"
echo "Go to /var/www and run ./start.sh (or ./start-ssl.sh if you've already set up ssl) to start the webserver! screen is recommended."


index.js
var http = require("http");
var https = require('https');
var url = require("url");
var fs = require("fs-extra");
var path = require("path");
var proc = require('child_process');
var URL = url.URL;
var par = "";
var conf = [];
var not_raw_download = [".html",".txt",".htm"];

var readconf = function(config) {
	var content = fs.readFileSync(config);
	conf = JSON.parse(content);
};

function parseCookies (request) {
	var list = {};
	var rc = request.headers.cookie;
	rc && rc.split(';').forEach(function( cookie ) {
		var parts = cookie.split('=');
		list[parts.shift().trim()] = decodeURI(parts.join('='));
	});
	return list;
}

var get = function(url, headers, answ) {
	var data = "";
	var tmp = url.split(":");
	var protocol = tmp[0];
	var options = new URL(url);
	if (headers == undefined) {headers = {};}
	options.headers = headers;
	if (protocol == "http"){
		var req = http.request(options, function(res) {
			res.setEncoding('utf8');
			res.on('data', function(body) {
				data = data+body;
			});
			res.on('end', function() {
				answ(data);
			});
		});
		req.on('error', function(err) {
			if(err) {
				console.log(err);
			}
		});
		req.write("");
		req.end();
	}
	if (protocol == "https") {
		var reqs = https.request(options, function(res) {
			res.setEncoding('utf8');
			res.on('data', function(body) {
				data = data+body;
			});
			res.on('end', function() {
				answ(data);
			});
		});
		reqs.on('error', function(err) {
			if(err) {
				console.log(err);
			}
		});
		reqs.write("");
		reqs.end();
	}
}

var post = function(url, stuff, headers, answ) {
	var data = "";
	var tmp = url.split(":");
	var protocol = tmp[0];
	var options = new URL(url);
	if (headers == undefined) {headers = {};}
	options.headers = headers;
	if (protocol == "http"){
		var req = http.request(options, function(res) {
			res.setEncoding('utf8');
			res.on('data', function(body) {
				data = data+body;
			});
			res.on('end', function() {
				answ(data);
			});
		});
		req.on('error', function(err) {
			if(err) {
				console.log(err);
			}
		});
		req.write(stuff);
		req.end();
	}
	if (protocol == "https") {
		var reqs = https.request(options, function(res) {
			res.setEncoding('utf8');
			res.on('data', function(body) {
				data = data+body;
			});
			res.on('end', function() {
				answ(data);
			});
		});
		reqs.on('error', function(err) {
			if(err) {
				console.log(err);
			}
		});
		reqs.write(stuff);
		reqs.end();
	}
};

var server = http.createServer(function(req, res) {
	if (process.argv.indexOf("--config") != -1) {
		readconf(process.argv[parseInt(process.argv.indexOf("--config"))+1]);
	} else if (process.argv.indexOf("-c") != -1) {
		readconf(process.argv[parseInt(process.argv.indexOf("-c"))+1]);
	} else {
		readconf("main.conf");
	}
	process.setgid("secureweb");
	process.setegid("secureweb");
	process.setuid("secureweb");
	process.seteuid("secureweb");
	var date = new Date();
	var hour = date.getHours();
	var minute = date.getMinutes();
	var second = date.getSeconds();
	var params = url.parse(req.url,true).pathname;
	var params2 = url.parse(req.url,true).query;
	var params3 = url.parse(req.url,true);
	if (params == "/") par = "/index.html"; else par = params;
	var dest = conf[req.headers.host.split(":")[0]];
	if (dest == undefined) {dest = conf["default"]}
	var ext = path.extname(par).toLowerCase();
	fs.appendFileSync("./log/log.txt", "[".concat(hour).concat(":").concat(minute).concat(":").concat(second).concat("] ").concat(req.connection.remoteAddress).concat(" : ").concat(req.method).concat(" : ").concat(req.headers.host).concat(req.url).concat("\n"));
	if (dest["type"] == "proxy") {
		if (req.method == "GET") {
			get(dest["location"].concat(params3.path), req.headers, function(answ){
				try {
					res.end(answ);
				} catch(err) {
					console.log(err);
				}
			});
		} else if (req.method == "POST") {
			var data = "";
			req.on('data', function(body){
				data = data+body;
			});
			req.on('end', function(){
				post(dest["location"].concat(params3.path), data, req.headers, function(answ){
					res.end(answ);
				});
			});
		} else {
			res.writeHead(405);
			res.end();
		}
	} else if (dest["type"] == "local") {
		fs.stat(dest["location"].concat(par), function(err, stats) {
			if (stats != undefined){
				if(stats.isDirectory()) {
					par = par + "/index.html";
				}
			}
			var exists = fs.existsSync(dest["location"].concat(par));
			if (exists == true) {exists = 1;} else {exists = 0;}
			if (exists == 1) {
				if (ext == ".sjs") {
					res.writeHead(200, {'Access-Control-Allow-Origin': '*'});
					if (req.method == "GET"){
						var parameters = {};
						try {parameters = url.parse(req.url,true).query} catch(err) {parameters = {}}
						var execopts = {
							maxBuffer: 100000000,
							env: {
								'PATH': process.env['PATH'],
								'COOKIES': req.headers.cookie
							}
						}
						var thing;
						for (thing in parameters) {
							execopts['env'][thing] = parameters[thing];
						}
						execopts['env']['PATH'] = process.env['PATH'];
						execopts['env']['PWD'] = dest['location'];
						execopts['env']['REQIP'] = req.connection.remoteAddress;
						execopts['env']['HEADERS'] = JSON.stringify(req.headers);
						proc.exec("node " + dest["location"].concat(par), execopts, function(err, stdout, stderr) {
							if (!err) {
								res.write(stdout);
							} else {
								console.log(err);
							}
							res.end();
						});
					} else if (req.method == "POST") {
						var data = "";
						req.on('data', function(body){
							data = data+body;
						});
						req.on('end', function(){
							var parameters = {};
							try {parameters = url.parse("?"+data,true).query;} catch(err) {parameters = {}}
							var execopts = {
								maxBuffer: 100000000,
								env: {
									'PATH': process.env['PATH'],
									'COOKIES': req.headers.cookie
								}
							}
							var thing;
							for (thing in parameters) {
								execopts['env'][thing] = parameters[thing];
							}
							execopts['env']['PATH'] = process.env['PATH'];
							execopts['env']['PWD'] = dest['location'];
							execopts['env']['REQIP'] = req.connection.remoteAddress;
							execopts['env']['HEADERS'] = JSON.stringify(req.headers);
							proc.exec("node " + dest["location"].concat(par), execopts, function(err, stdout, stderr) {
								if (!err) {
									res.write(stdout);
								} else {
									console.log(err);
								}
								res.end();
							});
						});
					}
				} else if (ext == ".jspt") {
					res.end("Under construction");
				} else if (ext == ".php" && dest["php_enabled"] != undefined) {
					if (req.method == "GET") {
						try {var parameters = req.url.split("?")[1].split("&").join(" ");} catch(err) {var parameters = ""}
						proc.exec("php " + dest["location"].concat(par) + ' "' + decodeURIComponent(parameters.replace(/Ã¯Â¿Â½/g, "'").replace(/Ã¯Â¿Â½/g, "'").replace(/%91/g, "'").replace(/%92/g, "'").replace(/%93/g, '"').replace(/%94/g, '"')).replace(/"/g, '\\"') +'"', function(err, stdout, stderr) {
							if (!err) {
								res.write(stdout);
								res.end();
							} else {
								console.log(err);
							}
						});
					} else if (req.method == "POST") {
						var data = "";
						req.on('data', function(body){
							data = data+body;
						});
						try {var parameters = data;} catch(err) {var parameters = ""}
						proc.exec("php " + dest["location"].concat(par) + ' "' + parameters.replace(/Ã¯Â¿Â½/g, "'").replace(/Ã¯Â¿Â½/g, "'").replace(/%91/g, "'").replace(/%92/g, "'").replace(/%93/g, '"').replace(/%94/g, '"').replace(/"/g, '\\"') +'"', function(err, stdout, stderr) {
							if (!err) {
								res.write(stdout);
								res.end();
							} else {
								console.log(err);
							}
						});	
					}  
				} else if (not_raw_download.indexOf(ext) == -1) {
					res.writeHead(200, {'Access-Control-Allow-Origin': '*'});
					var stats = fs.statSync(dest["location"].concat(par));
					fs.open(dest["location"].concat(par), 'r', function(status, fd) {
						if (status) {console.log(status.message); return;}
						var buffer = new Buffer(stats.size);
						fs.read(fd, buffer, 0, stats.size, 0, function(err, num) {
							res.write(buffer);
							if(num == stats.size) {res.end();};
						});
					});
				} else {
					res.writeHead(200, {'Access-Control-Allow-Origin': '*'});
					var data = fs.readFileSync(dest["location"].concat(par), 'utf8');
					res.end(data);
				}
			} else if (exists == 0) {
				var exists2 = fs.existsSync(dest["location"].concat("/404.html"));
				if (exists2) {
					res.writeHead(404);
					var data2 = fs.readFileSync(dest["location"].concat("/404.html"), 'utf8');
					res.end(data2);
				} else {
					res.writeHead(404);
					res.write("<!DOCTYPE html><html><head><title>Error 404. File not found!</title></head><body><h1>ERROR 404</h1>File ");
					res.write(par);
					res.write(" not found.<br/>------------------------<br/>JoshieHTTP/3.0.3_Linux<body></html>");
					res.end();
				}
			} else if (exists == 2) {
				res.writeHead(403);
				res.end("Access denied!");
			} else {
				var exists2 = fs.existsSync(dest["location"].concat("/500.html"));
				if (exists2) {
					var data2 = fs.readFileSync(dest["location"].concat("/500.html"), 'utf8');
					res.end(data2);
				} else {
					res.end("Error 500: internal server error");
				}
			}
		});
	}
});

if (process.argv.indexOf("--https") != -1 || process.argv.indexOf("-s") != -1) {
	if (process.argv.indexOf("--config") != -1) {
		readconf(process.argv[parseInt(process.argv.indexOf("--config"))+1]);
	} else if (process.argv.indexOf("-c") != -1) {
		readconf(process.argv[parseInt(process.argv.indexOf("-c"))+1]);
	} else {
		readconf("main.conf");
	}
	var options = {
		key: fs.readFileSync('ssl/key.pem'),
		cert: fs.readFileSync('ssl/cert.pem')
	};
	var sserver = https.createServer(options, function(req, res) {
		process.setgid("secureweb");
		process.setegid("secureweb");
		process.setuid("secureweb");
		process.seteuid("secureweb");
		var date = new Date();
		var hour = date.getHours();
		var minute = date.getMinutes();
		var second = date.getSeconds();
		var params = url.parse(req.url,true).pathname;
		var params2 = url.parse(req.url,true).query;
		var params3 = url.parse(req.url,true);
		if (params == "/") par = "/index.html"; else par = params;
		var dest = conf[req.headers.host.split(":")[0]];
		if (dest == undefined) {dest = conf["default"]}
		var ext = path.extname(par).toLowerCase();
		fs.appendFileSync("./log/log.txt", "[".concat(hour).concat(":").concat(minute).concat(":").concat(second).concat("] ").concat(req.connection.remoteAddress).concat(" : ").concat(req.method).concat(" : ").concat(req.headers.host).concat(req.url).concat("\n"));
		if (dest["type"] == "proxy") {
			if (req.method == "GET") {
				get(dest["location"].concat(params3.path), req.headers, function(answ){
					try {
						res.end(answ);
					} catch(err) {
						console.log(err);
					}
				});
			} else if (req.method == "POST") {
				var data = "";
				req.on('data', function(body){
					data = data+body;
				});
				req.on('end', function(){
					post(dest["location"], data, req.headers, function(answ){
						res.end(answ);
					});
				});
			} else {
				res.writeHead(405);
				res.end();
			}
		} else if (dest["type"] == "local") {
			fs.stat(dest["location"].concat(par), function(err, stats) {
				if (stats != undefined){
					if(stats.isDirectory()) {
						par = par + "/index.html";
					}
				}
				var exists = fs.existsSync(dest["location"].concat(par));
				if (exists == true) {exists = 1;} else {exists = 0;}
				if (exists == 1) {
					if (ext == ".sjs") {
						res.writeHead(200, {'Access-Control-Allow-Origin': '*'});
						if (req.method == "GET"){
							var parameters = {};
							try {parameters = url.parse(req.url,true).query} catch(err) {parameters = {}}
							var execopts = {
								maxBuffer: 100000000,
								env: {
									'PATH': process.env['PATH'],
									'COOKIES': req.headers.cookie
								}
							}
							var thing;
							for (thing in parameters) {
								execopts['env'][thing] = parameters[thing];
							}
							execopts['env']['PATH'] = process.env['PATH'];
							execopts['env']['PWD'] = dest['location'];
							execopts['env']['REQIP'] = req.connection.remoteAddress;
							execopts['env']['HEADERS'] = JSON.stringify(req.headers);
							proc.exec("node " + dest["location"].concat(par), execopts, function(err, stdout, stderr) {
								if (!err) {
									res.write(stdout);
								} else {
									console.log(err);
								}
								res.end();
							});
						} else if (req.method == "POST") {
							var data = "";
							req.on('data', function(body){
								data = data+body;
							});
							req.on('end', function(){
								var parameters = {};
								try {parameters = url.parse("?"+data,true).query;} catch(err) {parameters = {}}
								var execopts = {
									maxBuffer: 100000000,
									env: {
										'PATH': process.env['PATH'],
										'COOKIES': req.headers.cookie
									}
								}
								var thing;
								for (thing in parameters) {
									execopts['env'][thing] = parameters[thing];
								}
								execopts['env']['PATH'] = process.env['PATH'];
								execopts['env']['PWD'] = dest['location'];
								execopts['env']['REQIP'] = req.connection.remoteAddress;
								execopts['env']['HEADERS'] = JSON.stringify(req.headers);
								proc.exec("node " + dest["location"].concat(par), execopts, function(err, stdout, stderr) {
									if (!err) {
										res.write(stdout);
									} else {
										console.log(err);
									}
									res.end();
								});
							});
						}
					} else if (ext == ".jspt") {
						res.end("Under construction");
					} else if (ext == ".php" && dest["php_enabled"] != undefined) {
						if (req.method == "GET") {
							try {var parameters = req.url.split("?")[1].split("&").join(" ");} catch(err) {var parameters = ""}
							proc.exec("php " + dest["location"].concat(par) + ' "' + decodeURIComponent(parameters.replace(/Ã¯Â¿Â½/g, "'").replace(/Ã¯Â¿Â½/g, "'").replace(/%91/g, "'").replace(/%92/g, "'").replace(/%93/g, '"').replace(/%94/g, '"')).replace(/"/g, '\\"') +'"', function(err, stdout, stderr) {
								if (!err) {
									res.write(stdout);
									res.end();
								} else {
									console.log(err);
								}
							});
						} else if (req.method == "POST") {
							var data = "";
							req.on('data', function(body){
								data = data+body;
							});
							try {var parameters = data;} catch(err) {var parameters = ""}
							proc.exec("php " + dest["location"].concat(par) + ' "' + parameters.replace(/Ã¯Â¿Â½/g, "'").replace(/Ã¯Â¿Â½/g, "'").replace(/%91/g, "'").replace(/%92/g, "'").replace(/%93/g, '"').replace(/%94/g, '"').replace(/"/g, '\\"') +'"', function(err, stdout, stderr) {
								if (!err) {
									res.write(stdout);
									res.end();
								} else {
									console.log(err);
								}
							});	
						}  
					} else if (not_raw_download.indexOf(ext) == -1) {
						res.writeHead(200, {'Access-Control-Allow-Origin': '*'});
						var stats = fs.statSync(dest["location"].concat(par));
						fs.open(dest["location"].concat(par), 'r', function(status, fd) {
							if (status) {console.log(status.message); return;}
							var buffer = new Buffer(stats.size);
							fs.read(fd, buffer, 0, stats.size, 0, function(err, num) {
								res.write(buffer);
								if(num == stats.size) {res.end();};
							});
						});
					} else {
						res.writeHead(200, {'Access-Control-Allow-Origin': '*'});
						var data = fs.readFileSync(dest["location"].concat(par), 'utf8');
						res.end(data);
					}
				} else if (exists == 0) {
					var exists2 = fs.existsSync(dest["location"].concat("/404.html"));
					if (exists2) {
						res.writeHead(404);
						var data2 = fs.readFileSync(dest["location"].concat("/404.html"), 'utf8');
						res.end(data2);
					} else {
						res.writeHead(404);
						res.write("<!DOCTYPE html><html><head><title>Error 404. File not found!</title></head><body><h1>ERROR 404</h1>File ");
						res.write(par);
						res.write(" not found.<br/>------------------------<br/>JoshieHTTP/3.0.3_Linux<body></html>");
						res.end();
					}
				} else if (exists == 2) {
					res.writeHead(403);
					res.end("Access denied!");
				} else {
					var exists2 = fs.existsSync(dest["location"].concat("/500.html"));
					if (exists2) {
						var data2 = fs.readFileSync(dest["location"].concat("/500.html"), 'utf8');
						res.end(data2);
					} else {
						res.end("Error 500: internal server error");
					}
				}
			});
		}
	});
	if (process.argv.indexOf("--https") != -1) {
		sserver.listen(process.argv[parseInt(process.argv.indexOf("--https"))+1]);
	} else if (process.argv.indexOf("-s") != -1) {
		sserver.listen(process.argv[parseInt(process.argv.indexOf("-s"))+1]);
	}
}

if (process.argv.indexOf("--listen") != -1) {
	server.listen(process.argv[parseInt(process.argv.indexOf("--listen"))+1]);
} else if (process.argv.indexOf("-l") != -1) {
	server.listen(process.argv[parseInt(process.argv.indexOf("-l"))+1]);
} else {
	server.listen(8080);
}
if (process.argv.indexOf("--config") != -1) {
	readconf(process.argv[parseInt(process.argv.indexOf("--config"))+1]);
} else if (process.argv.indexOf("-c") != -1) {
	readconf(process.argv[parseInt(process.argv.indexOf("-c"))+1]);
} else {
	readconf("main.conf");
}

console.log("Started JoshieHTTPD/3.0.3_Linux");
